pkg/placement/declarations.go

```diff
@@ -0,0 +1,387 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apimachtypes "k8s.io/apimachinery/pkg/types"
+	k8ssets "k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	edgeapi "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
+)
+
+// This file contains declarations of the interfaces between the main parts
+// of the placement translator.  Those are as follows.
+//
+// - a WhereResolver that monitors the "where" resolutions in the
+//   SinglePlacementSlice API objects.
+// - a WhatResolver that monitors the EdgePlacement objects and resolves
+//   their "what" predicates.
+// - a SetBinder that keeps track of the bindings between resolved where
+//   and resolved what.
+// - a WorkloadProjector that maintains the customized workload copies
+//   in the mailbox workspaces.
+// - a PlacementProjector that maintains the TMC Placement objects that
+//   correspond to the EdgePlacement objects.
+
+// We are talking here about an assembly of several components, each
+// of which likely has private data protected by a mutex.
+// We thus must pay attention to avoiding deadlock.
+// We do that by declaring and respecting a partial order among mutexes.
+// When we say that mutex A precedes B in the locking order, this means that
+// it is forbidden for a goroutine to invoke `A.Lock()` while holding B locked.
+// Stated in a positive but little less precise way: a goroutine that is
+// going to lock both A and B must lock A first.
+
+// The particular locking order chosen here generally follows the pattern
+// that components that drive activity precede components that get driven,
+// so that this relationship can be synchronous.  For example, providers
+// of maps generally precede consumers of maps.
+
+// WhereResolver is responsible for keeping given consumers eventually
+// consistent with the resolution of the "where" predicate for each EdgePlacement
+// (identified by cluster and name).
+type WhereResolver DynamicMapProducer[edgeapi.ExternalName, ResolvedWhere]
+
+// WhatResolver is responsible for keeping its consumers eventually consistent
+// with the resolution of the "what" predicate of each EdgePlacement
+// (identified by cluster ane name).
+type WhatResolver DynamicMapProducer[edgeapi.ExternalName, WorkloadParts]
+
+// SetBinder is a component that is kept appraised of the "what" and "where"
+// resolutions and reorganizing and picking API versions to guide the
+// workload projector and the placement projector.
+// The implementation may atomize the resolved "where" using a SinglePlacementSliceSetReducer.
+// The implementation may use a BindingOrganizer to get from the atomized
+// "what" and "where" to the ProjectionMapProducer behavior.
+type SetBinder interface {
+	AsWhatConsumer() DynamicMapConsumer[edgeapi.ExternalName, WorkloadParts]
+	AsWhereConsumer() DynamicMapConsumer[edgeapi.ExternalName, ResolvedWhere]
+	ProjectionMapProducer
+}
+
+// WorkloadProjector is kept appraised of what goes where
```

Comment:
Is WorkloadProjector kept appraised of 'what goes where' or just 'what'?

Reply:
'what goes where'